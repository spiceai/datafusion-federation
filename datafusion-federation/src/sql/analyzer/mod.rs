mod collect_known_rewrites;
mod rewrite_table_scan;

pub use collect_known_rewrites::collect_known_rewrites;
pub use rewrite_table_scan::*;
