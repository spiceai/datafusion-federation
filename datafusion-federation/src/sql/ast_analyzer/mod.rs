mod rewrite_multi_part;
mod table_arg;

use datafusion::{common::error::Result, sql::sqlparser::ast};
pub use rewrite_multi_part::RewriteMultiTableReference;
pub use table_arg::TableArgReplace;

pub type AstAnalyzerRule = Box<dyn FnMut(ast::Statement) -> Result<ast::Statement>>;

#[derive(Default)]
pub struct AstAnalyzer {
    rules: Vec<AstAnalyzerRule>,
}

impl AstAnalyzer {
    pub fn new(rules: Vec<AstAnalyzerRule>) -> Self {
        Self { rules }
    }

    pub fn add_rule(&mut self, rule: AstAnalyzerRule) {
        self.rules.push(rule)
    }

    pub fn analyze(&mut self, mut statement: ast::Statement) -> Result<ast::Statement> {
        for rule in self.rules.iter_mut() {
            statement = rule(statement)?;
        }

        Ok(statement)
    }
}
