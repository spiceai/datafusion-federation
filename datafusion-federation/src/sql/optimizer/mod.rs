//! A collection of optimizer rules that benefit SQL federation.

mod push_down_filter;
pub use push_down_filter::PushDownFilterFederation;

mod optimize_projections;
pub use optimize_projections::OptimizeProjectionsFederation;
