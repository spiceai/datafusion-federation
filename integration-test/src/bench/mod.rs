use crate::validation::validate_tpch_query;
use anyhow::Result;
use datafusion::arrow::record_batch::RecordBatch;
use std::sync::Arc;

pub trait Benchmark {
    /// Name of the benchmark (e.g., "tpch", "tpcds")
    fn name(&self) -> &str;

    /// Database file name (e.g., "tpch_sf1.db", "tpcds_sf1.db")
    fn db_file_name(&self) -> String;

    /// Get all table names for this benchmark
    fn table_names(&self) -> Vec<String>;

    /// Get all queries for this benchmark
    fn queries(&self) -> Vec<Query>;

    fn validate(&self, query: &Query, batches: &[RecordBatch]) -> Result<()>;
}

pub struct Query {
    pub name: Arc<str>,
    pub sql: Arc<str>,
}

macro_rules! generate_tpch_queries {
     ( $( $i:tt ),* ) => {
         vec![
             $(
                 Query {
                     name: concat!("tpch_", stringify!($i)).into(),
                     sql: include_str!(concat!("../../queries/tpch/", stringify!($i), ".sql")).into(),
                 }
             ),*
         ]
     }
}

pub struct TpchBenchmark;

impl Benchmark for TpchBenchmark {
    fn name(&self) -> &str {
        "tpch"
    }

    fn db_file_name(&self) -> String {
        "tpch_sf1.db".to_string()
    }

    fn table_names(&self) -> Vec<String> {
        vec![
            "customer".to_string(),
            "orders".to_string(),
            "lineitem".to_string(),
            "part".to_string(),
            "supplier".to_string(),
            "partsupp".to_string(),
            "nation".to_string(),
            "region".to_string(),
        ]
    }

    fn queries(&self) -> Vec<Query> {
        generate_tpch_queries!(
            q1, q2, q3, q4, q5, q6, q7, q8, q9, q10, q11, q12, q13, q14, q16, q17, q18, q19, q20,
            q21, q22
        )
    }

    fn validate(&self, query: &Query, batches: &[RecordBatch]) -> Result<()> {
        validate_tpch_query(query, batches)
    }
}
