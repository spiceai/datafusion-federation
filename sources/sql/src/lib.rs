use core::fmt;
use std::{any::Any, collections::HashMap, sync::Arc, vec};

use async_trait::async_trait;
use datafusion::{
    arrow::datatypes::{Schema, SchemaRef},
    config::ConfigOptions,
    error::Result,
    execution::{context::SessionState, TaskContext},
    logical_expr::{Extension, LogicalPlan},
    optimizer::analyzer::{Analyzer, AnalyzerRule},
    physical_expr::EquivalenceProperties,
    physical_plan::{
        DisplayAs, DisplayFormatType, ExecutionMode, ExecutionPlan, Partitioning, PlanProperties,
        SendableRecordBatchStream,
    },
    sql::{
        unparser::{plan_to_sql, Unparser},
        TableReference,
    },
};
use datafusion_federation::{
    schema_cast,
    table_reference::{MultiPartTableReference, MultiTableReference},
    FederatedPlanNode, FederationPlanner, FederationProvider,
};

mod schema;
pub use schema::*;

mod executor;
pub use executor::*;

mod rewrite;

// SQLFederationProvider provides federation to SQL DMBSs.
pub struct SQLFederationProvider {
    analyzer: Arc<Analyzer>,
    executor: Arc<dyn SQLExecutor>,
}

impl SQLFederationProvider {
    pub fn new(executor: Arc<dyn SQLExecutor>) -> Self {
        Self {
            analyzer: Arc::new(Analyzer::with_rules(vec![Arc::new(
                SQLFederationAnalyzerRule::new(Arc::clone(&executor)),
            )])),
            executor,
        }
    }
}

impl FederationProvider for SQLFederationProvider {
    fn name(&self) -> &str {
        "sql_federation_provider"
    }

    fn compute_context(&self) -> Option<String> {
        self.executor.compute_context()
    }

    fn analyzer(&self) -> Option<Arc<Analyzer>> {
        Some(Arc::clone(&self.analyzer))
    }
}

struct SQLFederationAnalyzerRule {
    planner: Arc<dyn FederationPlanner>,
}

impl std::fmt::Debug for SQLFederationAnalyzerRule {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("SQLFederationAnalyzerRule").finish()
    }
}

impl SQLFederationAnalyzerRule {
    pub fn new(executor: Arc<dyn SQLExecutor>) -> Self {
        Self {
            planner: Arc::new(SQLFederationPlanner::new(Arc::clone(&executor))),
        }
    }
}

impl AnalyzerRule for SQLFederationAnalyzerRule {
    fn analyze(&self, plan: LogicalPlan, _config: &ConfigOptions) -> Result<LogicalPlan> {
        let fed_plan = FederatedPlanNode::new(plan.clone(), Arc::clone(&self.planner));
        let ext_node = Extension {
            node: Arc::new(fed_plan),
        };
        Ok(LogicalPlan::Extension(ext_node))
    }

    /// A human readable name for this analyzer rule
    fn name(&self) -> &str {
        "federate_sql"
    }
}

struct SQLFederationPlanner {
    executor: Arc<dyn SQLExecutor>,
}

impl SQLFederationPlanner {
    pub fn new(executor: Arc<dyn SQLExecutor>) -> Self {
        Self { executor }
    }
}

#[async_trait]
impl FederationPlanner for SQLFederationPlanner {
    async fn plan_federation(
        &self,
        node: &FederatedPlanNode,
        _session_state: &SessionState,
    ) -> Result<Arc<dyn ExecutionPlan>> {
        let schema = Arc::new(node.plan().schema().as_arrow().clone());
        let input = Arc::new(VirtualExecutionPlan::new(
            node.plan().clone(),
            Arc::clone(&self.executor),
        ));
        let schema_cast_exec = schema_cast::SchemaCastScanExec::new(input, schema);
        Ok(Arc::new(schema_cast_exec))
    }
}

#[derive(Debug, Clone)]
struct VirtualExecutionPlan {
    plan: LogicalPlan,
    executor: Arc<dyn SQLExecutor>,
    props: PlanProperties,
}

impl VirtualExecutionPlan {
    pub fn new(plan: LogicalPlan, executor: Arc<dyn SQLExecutor>) -> Self {
        let schema: Schema = plan.schema().as_ref().into();
        let props = PlanProperties::new(
            EquivalenceProperties::new(Arc::new(schema)),
            Partitioning::UnknownPartitioning(1),
            ExecutionMode::Bounded,
        );
        Self {
            plan,
            executor,
            props,
        }
    }

    fn schema(&self) -> SchemaRef {
        let df_schema = self.plan.schema().as_ref();
        Arc::new(Schema::from(df_schema))
    }

    fn sql(&self) -> Result<String> {
        // Find all table scans, recover the SQLTableSource, find the remote table name and replace the name of the TableScan table.
        let mut known_rewrites = HashMap::new();
        let subquery_uses_partial_path = self.executor.subquery_use_partial_path();
        let rewritten_plan = rewrite::plan::rewrite_table_scans(
            &self.plan,
            &mut known_rewrites,
            subquery_uses_partial_path,
            &mut None,
        )?;
        let mut ast =
            Unparser::new(self.executor.dialect().as_ref()).plan_to_sql(&rewritten_plan)?;

        // If there are any MultiPartTableReference, rewrite the AST to use the original table names.
        let multi_table_reference_rewrites = known_rewrites
            .into_iter()
            .filter_map(|(table_ref, rewrite)| match rewrite {
                MultiPartTableReference::Multi(rewrite) => Some((table_ref, rewrite)),
                _ => None,
            })
            .collect::<HashMap<TableReference, MultiTableReference>>();
        tracing::trace!("multi_table_reference_rewrites: {multi_table_reference_rewrites:?}");
        if !multi_table_reference_rewrites.is_empty() {
            rewrite::ast::rewrite_multi_part_statement(&mut ast, &multi_table_reference_rewrites);
        }

        if let Some(analyzer) = self.executor.ast_analyzer() {
            ast = analyzer(ast)?;
        }

        Ok(format!("{ast}"))
    }
}

impl DisplayAs for VirtualExecutionPlan {
    fn fmt_as(&self, _t: DisplayFormatType, f: &mut fmt::Formatter) -> std::fmt::Result {
        write!(f, "VirtualExecutionPlan")?;
        let Ok(ast) = plan_to_sql(&self.plan) else {
            return Ok(());
        };
        write!(f, " name={}", self.executor.name())?;
        if let Some(ctx) = self.executor.compute_context() {
            write!(f, " compute_context={ctx}")?;
        };

        write!(f, " sql={ast}")?;
        if let Ok(query) = self.sql() {
            write!(f, " rewritten_sql={query}")?;
        };

        Ok(())
    }
}

impl ExecutionPlan for VirtualExecutionPlan {
    fn name(&self) -> &str {
        "VirtualExecutionPlan"
    }

    fn as_any(&self) -> &dyn Any {
        self
    }

    fn schema(&self) -> SchemaRef {
        self.schema()
    }

    fn children(&self) -> Vec<&Arc<dyn ExecutionPlan>> {
        vec![]
    }

    fn with_new_children(
        self: Arc<Self>,
        _: Vec<Arc<dyn ExecutionPlan>>,
    ) -> Result<Arc<dyn ExecutionPlan>> {
        Ok(self)
    }

    fn execute(
        &self,
        _partition: usize,
        _context: Arc<TaskContext>,
    ) -> Result<SendableRecordBatchStream> {
        self.executor.execute(self.sql()?.as_str(), self.schema())
    }

    fn properties(&self) -> &PlanProperties {
        &self.props
    }
}
