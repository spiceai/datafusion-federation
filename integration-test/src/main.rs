mod bench;
mod validation;

use anyhow::{anyhow, Result};
use bench::{Benchmark, Query};
use std::path::{Path, PathBuf};
use std::sync::Arc;

use datafusion::arrow::util::pretty::pretty_format_batches;
use datafusion::sql::TableReference;
use datafusion_federation::{sql::federation_analyzer_rule, FederatedQueryPlanner};
use datafusion_table_providers::{
    duckdb::DuckDBTableFactory, sql::db_connection_pool::duckdbpool::DuckDbConnectionPool,
};
use duckdb::{AccessMode, Connection};

use datafusion::{
    execution::{context::SessionContext, session_state::SessionStateBuilder},
    optimizer::{
        analyzer::{
            resolve_grouping_function::ResolveGroupingFunction, type_coercion::TypeCoercion,
        },
        AnalyzerRule,
    },
};

pub fn get_analyzer_rules() -> Vec<Arc<dyn AnalyzerRule + Send + Sync>> {
    vec![
        Arc::new(federation_analyzer_rule()),
        Arc::new(ResolveGroupingFunction::new()),
        Arc::new(TypeCoercion::new()),
    ]
}

fn generate_tpch_database(db_path: &Path) -> Result<()> {
    println!("Generating TPC-H database at: {}", db_path.display());

    let db_path_str = db_path
        .to_str()
        .ok_or_else(|| anyhow!("Failed to convert db_path to string"))?;

    let conn = Connection::open(db_path_str)
        .map_err(|e| anyhow!("Failed to open DuckDB connection: {}", e))?;

    // Install TPC-H extension
    conn.execute("INSTALL tpch;", [])
        .map_err(|e| anyhow!("Failed to install tpch extension: {}", e))?;

    // Load TPC-H extension
    conn.execute("LOAD tpch;", [])
        .map_err(|e| anyhow!("Failed to load tpch extension: {}", e))?;

    // Generate TPC-H data with scale factor 1
    conn.execute("CALL dbgen(sf=1);", [])
        .map_err(|e| anyhow!("Failed to generate tpch data: {}", e))?;

    println!("TPC-H database generated successfully");
    Ok(())
}

fn get_duckdb_table_factory(db_name: String) -> Result<DuckDBTableFactory> {
    let db_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .parent()
        .ok_or_else(|| anyhow!("Failed to get parent directory"))?
        .join(db_name);

    // Check if database exists, if not generate it
    if !db_path.exists() {
        generate_tpch_database(&db_path)?;
    }

    let duckdb_pool = Arc::new(
        DuckDbConnectionPool::new_file(
            db_path
                .to_str()
                .ok_or_else(|| anyhow!("Failed to convert db_path to string"))?,
            &AccessMode::ReadOnly,
        )
        .map_err(|e| anyhow!("Failed to create DuckDB connection pool: {}", e))?,
    );

    Ok(DuckDBTableFactory::new(duckdb_pool))
}

#[tokio::main]
async fn main() {
    let benchmark: Arc<dyn Benchmark> = Arc::new(bench::TpchBenchmark);

    let duckdb_table_factory = get_duckdb_table_factory(benchmark.db_file_name())
        .expect("unable to create DuckDB connection pool");

    let state = SessionStateBuilder::new()
        .with_query_planner(Arc::new(FederatedQueryPlanner::new()))
        .with_analyzer_rules(get_analyzer_rules())
        .with_default_features()
        .build();
    let ctx = SessionContext::new_with_state(state);

    let _ = register_federated_duckdb_tables(&ctx, benchmark.table_names(), &duckdb_table_factory)
        .await;

    let test_queries = benchmark.queries();

    for query in test_queries {
        run_test_query(&ctx, Arc::clone(&benchmark), &query)
            .await
            .unwrap_or_else(|err| {
                panic!(
                    "Failed to run {} query {}: {}",
                    benchmark.name(),
                    query.name,
                    err
                )
            });
    }
}

async fn run_test_query(
    ctx: &SessionContext,
    benchmark: Arc<dyn Benchmark>,
    query: &Query,
) -> Result<()> {
    let df = ctx.sql(&query.sql).await?;

    let plan = df.clone().explain(false, false)?.collect().await?;
    let plan_display = pretty_format_batches(&plan)?.to_string();

    insta::with_settings!({
        description => format!("Federated Query Explain"),
        snapshot_path => "../snapshots/explain",
        filters => vec![
            (r"compute_context=.*/([^/]+\.db)", "compute_context=$1")
        ],
    }, {
        insta::assert_snapshot!(format!("{}_{}_explain", benchmark.name(), query.name), plan_display.clone());
    });

    let result = df.collect().await.inspect_err(|_| {
        println!("{plan_display}");
    })?;
    benchmark.validate(query, &result)?;

    Ok(())
}

async fn register_federated_duckdb_tables(
    ctx: &SessionContext,
    table_names: Vec<String>,
    duckdb_table_factory: &DuckDBTableFactory,
) -> Result<()> {
    for table_name in table_names {
        ctx.register_table(
            &table_name,
            duckdb_table_factory
                .table_provider(TableReference::bare(table_name.as_str()))
                .await
                .map_err(|e| anyhow!("Failed to register duckdb table: {}", e))?,
        )?;
    }

    Ok(())
}
