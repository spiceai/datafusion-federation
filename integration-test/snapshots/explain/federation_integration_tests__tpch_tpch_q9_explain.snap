---
source: integration-test/src/main.rs
description: Federated Query Explain
expression: plan_display.clone()
---
+---------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      |
+---------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| logical_plan  | Federated                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 |
|               |  Sort: profit.nation ASC NULLS LAST, profit.o_year DESC NULLS FIRST                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       |
|               |   Projection: profit.nation, profit.o_year, sum(profit.amount) AS sum_profit                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              |
|               |     Aggregate: groupBy=[[profit.nation, profit.o_year]], aggr=[[sum(profit.amount)]]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      |
|               |       SubqueryAlias: profit                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               |
|               |         Projection: nation.n_name AS nation, date_part(Utf8("YEAR"), orders.o_orderdate) AS o_year, lineitem.l_extendedprice * (Int64(1) - lineitem.l_discount) - partsupp.ps_supplycost * lineitem.l_quantity AS amount                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  |
|               |           Inner Join:  Filter: supplier.s_nationkey = nation.n_nationkey                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  |
|               |             Inner Join:  Filter: orders.o_orderkey = lineitem.l_orderkey                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  |
|               |               Inner Join:  Filter: partsupp.ps_suppkey = lineitem.l_suppkey AND partsupp.ps_partkey = lineitem.l_partkey                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  |
|               |                 Inner Join:  Filter: supplier.s_suppkey = lineitem.l_suppkey AND part.p_partkey = lineitem.l_partkey                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      |
|               |                   Cross Join:                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             |
|               |                     TableScan: part projection=[p_partkey], full_filters=[part.p_name LIKE Utf8("%green%")]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               |
|               |                     TableScan: supplier projection=[s_suppkey, s_nationkey]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               |
|               |                   TableScan: lineitem projection=[l_orderkey, l_partkey, l_suppkey, l_quantity, l_extendedprice, l_discount]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              |
|               |                 TableScan: partsupp projection=[ps_partkey, ps_suppkey, ps_supplycost]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
|               |               TableScan: orders projection=[o_orderkey, o_orderdate]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      |
|               |             TableScan: nation projection=[n_nationkey, n_name]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            |
| physical_plan | SchemaCastScanExec                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        |
|               |   VirtualExecutionPlan name=duckdb compute_context=tpch_sf1.db initial_sql=SELECT "profit"."nation", "profit"."o_year", sum("profit"."amount") AS "sum_profit" FROM (SELECT "nation"."n_name" AS "nation", date_part('YEAR', "orders"."o_orderdate") AS "o_year", (("lineitem"."l_extendedprice" * (1 - "lineitem"."l_discount")) - ("partsupp"."ps_supplycost" * "lineitem"."l_quantity")) AS "amount" FROM "part" INNER JOIN "supplier" ON "part"."p_name" LIKE '%green%' INNER JOIN "lineitem" ON (("supplier"."s_suppkey" = "lineitem"."l_suppkey") AND ("part"."p_partkey" = "lineitem"."l_partkey")) INNER JOIN "partsupp" ON (("partsupp"."ps_suppkey" = "lineitem"."l_suppkey") AND ("partsupp"."ps_partkey" = "lineitem"."l_partkey")) INNER JOIN "orders" ON ("orders"."o_orderkey" = "lineitem"."l_orderkey") INNER JOIN "nation" ON ("supplier"."s_nationkey" = "nation"."n_nationkey")) AS "profit" GROUP BY "profit"."nation", "profit"."o_year" ORDER BY "profit"."nation" ASC NULLS LAST, "profit"."o_year" DESC NULLS FIRST |
|               |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           |
+---------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
